submariner-operator/submariner-metrics-redesign.md

```diff
@@ -0,0 +1,146 @@
+* A new DaemonSet `submariner-metrics` will be created. It will use the NodeSelector `status.submariner.io/gateway=active`
+to make sure it always runs on the active gateway node.
+* This DaemonSet will use `schedulerName: default-scheduler` to make sure that metrics proxy pod moves with the active gateway
+label.
+* DaemonSet will also use toleration to make sure it can always be scheduled on gateway node, irrespective of any taints.
+* `submariner-metrics` Pod will use `netcat` to act as proxy and forward all incoming requests on metrics port to the
+metrics port on the node IP.
+* The proxy pod will run two containers if globlanet is enabled, one each for gateway and globalnet.
+* Existing metrics service for gateway and globalnet will be modified to use the new metrics proxy pod. This will be
+done by using same label selector as metrics DaemonSet.
+
+### Submariner Metrics DaemonSet
+
+```yaml
+---
+apiVersion: apps/v1
+kind: DaemonSet
+metadata:
+  name: submariner-metrics
+spec:
+  selector:
+    matchLabels:
+      app: submariner-metrics
+  template:
+    metadata:
+      labels:
+        app: submariner-metrics
+    spec:
+      nodeSelector:
+        "status.submariner.io/gateway": "active"
+      containers:
+        - name: gateway-metrics-proxy
+          image: quay.io/submariner/nettest:devel
+          imagePullPolicy: IfNotPresent
+          env:
+          - name: NODE_IP
+            valueFrom:
+              fieldRef:
+                fieldPath: status.hostIP
+          - name: SUBM_GW_METRICS_PORT
+            value: "8080"
+          command: ["/bin/sh", "-c", "nc -v -lk -p 8080 -e nc $(NODE_IP) $(SUBM_GW_METRICS_PORT)"]
+        - name: globalnet-metrics-proxy
+          image: quay.io/submariner/nettest:devel
+          imagePullPolicy: IfNotPresent
+          env:
+          - name: NODE_IP
+            valueFrom:
+              fieldRef:
+                fieldPath: status.hostIP
+          - name: SUBM_GN_METRICS_PORT
+            value: "8081"
+          command: ["/bin/sh", "-c", "nc -v -lk -p 8081 -e nc $(NODE_IP) $(SUBM_GN_METRICS_PORT)"]
+      tolerations:
+      - operator: Exists
+      restartPolicy: Always
+      schedulerName: default-scheduler
+```
+
+### Changes to Submariner Gateway Metrics service
+
+```diff
+metadata:
+  name: submariner-gateway-metrics
+  labels:
+-   app: submariner-gateway
++   app: submariner-metrics
+...
+spec:
+  selector:
+-   app: submariner-gateway
++   app: submariner-metrics
+```
+
+### Changes to Submariner Globalnet Metrics service
+
+```diff
+metadata:
+  name: submariner-globalnet-metrics
+  labels:
+-   app: submariner-globalnet
++   app: submariner-metrics
+...
+spec:
+  selector:
+-   app: submariner-globalnet
++   app: submariner-metrics
+```
+
+### Pros
+
+* No user facing changes to existing services
+* Only one extra pod created for both gateway and globalnet
+
+### Cons
+
+* Requires changes in route-agent code to add the new label.
```

Comment:
Not if we use pod selector

Reply:
As clarified above, we can remove the changes associated with Route-Agent